### ISSUE TITLE
Nav Unification - clicking on sections causes UI to "jump" and triggers other sections to close.

### ISSUE THREAD

**Issue by getdave** [Automattic/wp-calypso#46091]
Open a menu section. Then click on another menu section. This will sometimes cause the other menu sections to toggle closed.

For example in the screencapture below:

* `Pages` is toggled open.
* `Tools` is clicked.
* `Tools` opens and `Pages` closes. This is not expected.

![Screen Capture on 2020-10-01 at 11-29-29](https://user-images.githubusercontent.com/444434/94798464-8808d980-03d9-11eb-99aa-40371a4a2fce.gif)

I believe this is due to the sidebar completely re-rendering when the route `path` prop changes. Given that `path` is passed down the React tree several levels deep any change to this will cause the _entire_ nav to render (including the Site Switcher and Notices).

![Screen Shot 2020-10-01 at 12 45 23](https://user-images.githubusercontent.com/444434/94805409-3ca7f880-03e4-11eb-91ee-1d31a968423b.png)

Here's a flamegraph from the React profiler showing the render profile when the `Pages` menu item is clicked.

**Comment by getdave** [comment]
# Update

As I may not be able to see this over the line (see p242cR-keI-p2) here's a proposed route forward for whoever picks this up.

The key problems are:

1. Optimising the `messagePath` and `jitms` props going to `SiteNotice` and `JITM` components. 
2. Optimise and improve the state management of JITMs in Redux to avoid the re-rendering.
3. Merge the above _without_ breaking existing implementations.

We have x3 PRs all try to address the `messagePath` prop on the component side. I list them below in the order of priority based on my preferred implementation:

1. https://github.com/Automattic/wp-calypso/pull/46713
2. https://github.com/Automattic/wp-calypso/pull/46688
3. https://github.com/Automattic/wp-calypso/pull/46676

I would personally go with `#1` as it seems like the best overall approach. Given the potential for this to affect JITMs showing across Calypso I would recommend we get input from at least 1 external reviewer. Ideally @sixhours if possible as she has worked on this previously.

Following on from that we need to merge whichever approach we take for `messagePath` and then rebase https://github.com/Automattic/wp-calypso/pull/46677 which handles optimisation of the JITMs in state. 

Note: as https://github.com/Automattic/wp-calypso/pull/46677 was based on https://github.com/Automattic/wp-calypso/pull/46676 we'll need to remove those changes first before rebasing.

Once that is done, then we need some final external reviews on https://github.com/Automattic/wp-calypso/pull/46677 in order to merge and deploy. With that done the "jumping" behaviour should be removed (from both unified nav _and_ the current Calypso nav!).

cc @obenland @davemart-in

**Comment by mreishus** [comment]
Update:

* #46713 has been approved (including external); I will merge it on Monday (2020-10-26)
* #46688 and #46676 are superseded by #46713
* #46677 is still in review and attacks the problem from a different angle; still in review, but is looking good.

Either one of (#46713, #46677) is sufficient to fix the jumpiness, but by merging both we'll fix it and get a small performance boost, too.
